### PR TITLE
Add support for halting execution plans

### PR DESCRIPTION
**Pull request by adamruzicka**
Up until now, Dynflow only had support for clean cancellation. Upon cancellation
a cancel event was sent to all cancellable suspended and running steps and it
was the action's responsibility to cancel itself. However, there were actions
which just weren't cancellable at all.

This commit adds support for halting execution plans. The actions are completely
unaware of this, on halt dynflow just destroys its internal state about a given
execution plan and turns it to stopped state. It does not remove running steps
from workers as we currently don't have any real means of doing so. Any pending
events will be rejected. In other words, halting an execution plan ensure it
will not run any further.

### CODE REVIEW

**Comment by ezr-ondrej** [comment]
This sounds like a nice feature, but I do not feel comfortable enough to ellaborate on the consequences here...

**Comment by ezr-ondrej** [comment]
@adamruzicka if you'd like me to give it another try, I can do a review here. I've improved my code reading skills and I've learned YOLO, so I'd be happy to merge after a proper review

**Comment by adamruzicka** [comment]
I was about to merge this myself in the upcoming weeks, but I'll never say no to another pair of eyes on a pr so if you could, I'd be grateful for that

**Review by ezr-ondrej — APPROVED** [review]
Overall looks pretty well designed :)
There is one too many `halted` flags to my taste and I'm unsure if the clean-up is all correct, but looks like it! :)
+1 from me FWIW.

**Comment by adamruzicka** [comment]
Thank you @ezr-ondrej !

**Pull request merged**
